src/number_theory/fermat4.lean

```diff
@@ -0,0 +1,335 @@
+  -- a ^ 2 = m ^ 2 - n ^ 2, b ^ 2 = 2 * m * n and c = m ^ 2 + n ^ 2
+  -- first the formula:
+  have ht : pythagorean_triple (a ^ 2) (b ^ 2) c,
+  { calc ((a ^ 2) * (a ^ 2)) + ((b ^ 2) * (b ^ 2))
+        = a ^ 4 + b ^ 4 : by ring
+    ... = c ^ 2 : by rw h.1.2.2
+    ... = c * c : by rw pow_two
+  },
+  -- coprime requirement:
+  have h2 : int.gcd (a ^ 2) (b ^ 2) = 1,
+  { change nat.coprime (int.nat_abs (a ^ 2)) (int.nat_abs (b ^ 2)),
+    rw [pow_two, int.nat_abs_mul],
+    apply nat.coprime.mul;
+    { rw [pow_two, int.nat_abs_mul], apply nat.coprime.mul_right;
+      apply (coprime_of_minimal h) } },
+  -- in order to reduce the possibilities we get from the classification of pythagorean triples
+  -- it helps if we know the parity of a ^ 2 (and the sign of c):
+  have ha22 : a ^ 2 % 2 = 1, { rw [pow_two, int.mul_mod, ha2], norm_num },
+  obtain ⟨m, n, ht1, ht2, ht3, ht4, ht5, ht6⟩ :=
+    pythagorean_triple.coprime_classification' ht h2 ha22 hc,
+  -- Now a, n, m form a pythagorean triple and so we can obtain r and s such that
+  -- a = r ^ 2 - s ^ 2, n = 2 * r * s and m = r ^ 2 + s ^ 2
+  -- formula:
+  have htt : pythagorean_triple a n m,
+  { delta pythagorean_triple,
+    rw [← pow_two, ← pow_two, ← pow_two], exact (add_eq_of_eq_sub ht1) },
+  -- a and n are coprime, because a ^ 2 = m ^ 2 - n ^ 2 and m and n are coprime.
+  have h3 : int.gcd a n = 1,
+  { have ha : (int.gcd a n : ℤ) ^ 2 ∣ a ^ 2,
+    { exact (int.pow_dvd_pow_iff (dec_trivial : 0 < 2)).mpr (int.gcd_dvd_left a n),},
+    have hn : (int.gcd a n : ℤ) ^ 2 ∣ n ^ 2,
+    { exact (int.pow_dvd_pow_iff (dec_trivial : 0 < 2)).mpr (int.gcd_dvd_right a n),},
+    have hm : (int.gcd a n : ℤ) ^ 2 ∣ m ^ 2,
+    { rw [← (add_eq_of_eq_sub ht1)], exact dvd_add ha hn },
+    replace hm : (int.gcd a n : ℤ) ∣ m, { exact (int.pow_dvd_pow_iff (dec_trivial : 0 < 2)).mp hm},
+    apply nat.eq_one_of_dvd_one,
+    rw [←ht4], apply int.coe_nat_dvd.mp, exact int.dvd_gcd hm (int.gcd_dvd_right a n) },
+  -- m is positive because b is non-zero and b ^ 2 = 2 * m * n and we already have 0 ≤ m.
+  have hb20 : b ^ 2 ≠ 0, { exact mt pow_eq_zero h.1.2.1 },
+  have h4 : 0 < m,
+  { apply lt_of_le_of_ne ht6,
+    by_contradiction hm0,
+    revert hb20,
+    rw [ht2, ← (not_not.mp hm0)], simp },
+  obtain ⟨r, s, htt1, htt2, htt3, htt4, htt5, htt6⟩ :=
+    pythagorean_triple.coprime_classification' htt h3 ha2 h4,
+  -- Now use the fact that (b / 2) ^ 2 = m * r * s, and m, r and s are pairwise coprime to obtain
+  -- i, j and k such that m = i ^ 2, r = j ^ 2 and s = k ^ 2.
+  -- m and r * s are coprime because m = r ^ 2 + s ^ 2 and r and s are coprime.
+  have hcp : int.gcd m (r * s) = 1, exact int.coprime_of_sqr_sum' htt3 htt4,
+  -- b is even because b ^ 2 = 2 * m * n.
+  have hb2 : 2 ∣ b,
+  { apply (or_self _).mp (int.prime.dvd_mul' (by norm_num : nat.prime 2) _),
+    rw ← pow_two, rw [ht2, mul_assoc], exact dvd_mul_right 2 (m * n) },
+  have hs : (b / 2) ^ 2 = m * (r * s),
```

Comment:
I would have been tempted to do cases on hb2 to get b = 2 * b' and then use b' instead of b/2 -- division on integers is annoying to use.

Reply:
@paulvanwamelen What do you think of this suggestion?

Reply:
I thought this was just a comment, not a suggestion. I'll change my code.